### ISSUE TITLE
Options page not found when submitting Expert Settings

### ISSUE THREAD

**Issue by blarghmatey** [podlove/podlove-publisher#862]
For any of the 'Expert Settings' tabs, when I click the submit button I receive an 'Error: options page not found' message.

System report:
```
Website             https://www.podcastinit.com
PHP Version         7.0.10-1~dotdeb+8.1
WordPress Version   4.6
WordPress Theme     Twenty Sixteen v1.3
Publisher Version   2.3.13
Web Player Version  2.0.19
Twig Version        1.23.1
open_basedir        ok
curl Version        7.38.0
iconv               available
simplexml           ok
max_execution_time  400
upload_max_filesize 200M
memory_limit        256M
disable_classes     
disable_functions   pcntl_alarm,pcntl_fork,pcntl_waitpid,pcntl_wait,pcntl_wifexited,pcntl_wifstopped,pcntl_wifsignaled,pcntl_wifcontinued,pcntl_wexitstatus,pcntl_wtermsig,pcntl_wstopsig,pcntl_signal,pcntl_signal_dispatch,pcntl_get_last_error,pcntl_strerror,pcntl_sigprocmask,pcntl_sigwaitinfo,pcntl_sigtimedwait,pcntl_exec,pcntl_getpriority,pcntl_setpriority,
permalinks          ok (/index.php/%postname%/)
podlove_permalinks  ok
podcast_settings    ok
web_player          ok
podlove_cache       on
assets              
  - mp3    audio/mpeg https://www.podcastinit.com/index.php/feed/episodes/
  - ogg    audio/ogg  https://www.podcastinit.com/index.php/feed/ogg-episodes/

0 errors
0 notices
Nice, Everything looks fine!
```

**Comment by timpritlove** [comment]
GitHub issues are for code errors only so please file a support request at <https://community.podlove.org/c/podlove-publisher>. Looks like an issue with WordPress at first sight.

**Comment by blarghmatey** [comment]
@timpritlove I think it might be related to the 'whitelist_options' filter as outlined at https://codex.wordpress.org/Function_Reference/register_setting

I am able to replicate this on a Vagrant vm with a fresh install of Wordpress and the Podlove Publisher as the only installed plugin. Part of the issue might be due to the fact that I'm using PHP 7, I'm not sure what the current state of compatibility happens to be for Podlove with that version.

```
Website             http://www.podcastinit.com
PHP Version         7.0.10-1~dotdeb+8.1
WordPress Version   4.6
WordPress Theme     Twenty Sixteen v1.3
Publisher Version   2.3.13
Web Player Version  2.0.19
Twig Version        1.23.1
open_basedir        ok
curl Version        7.38.0
iconv               available
simplexml           ok
max_execution_time  30
upload_max_filesize 2M
memory_limit        256M
disable_classes     
disable_functions   pcntl_alarm,pcntl_fork,pcntl_waitpid,pcntl_wait,pcntl_wifexited,pcntl_wifstopped,pcntl_wifsignaled,pcntl_wifcontinued,pcntl_wexitstatus,pcntl_wtermsig,pcntl_wstopsig,pcntl_signal,pcntl_signal_dispatch,pcntl_get_last_error,pcntl_strerror,pcntl_sigprocmask,pcntl_sigwaitinfo,pcntl_sigtimedwait,pcntl_exec,pcntl_getpriority,pcntl_setpriority,
permalinks          "non-pretty" Permalinks: Please change permalink structure
podlove_permalinks  ok
podcast_settings    Your podcast needs a title.
web_player          You need to assign at least one asset to the web player.
podlove_cache       on
assets              
  - 

%s ERRORS: 
- You are using the default WordPress permalink structure. This may cause problems with some podcast clients. Go to http://www.podcastinit.com/wp-admin/options-permalink.php and set it to anything but default (for example "Post name").
- Your podcast needs a title.
- You need to assign at least one asset to the web player.
0 notices
```

**Comment by timpritlove** [comment]
Does the user you are using have _Administrator_ privileges?

**Comment by blarghmatey** [comment]
Yes, I'm the only user on the system and my account was created as part of the Wordpress installation process. I am able to modify every other setting in the Podlove plugin, it is only in the expert settings panel that I am receiving this error.

**Comment by eteubert** [comment]
We had the same issue in https://github.com/podlove/podlove-publisher/issues/739 and I found/fixed an issue in 2ca83d98dc53b4150931c36869a4316bea1309fc.

I don't think it's PHP7 related. What exact vagrant box are you using?

**Comment by blarghmatey** [comment]
I am using the debian/jessie64 vagrant bospx and Debian 8 on my Linode instance.

**Comment by blarghmatey** [comment]
After some more research it appears that the issue at hand might be related to this: http://stackoverflow.com/a/20247875/3022059

It seems that the settings that are being set are not in the global whitelist. Below is a sample of the form data that is being submitted.

```
podlove_tab:web-player
option_page:podlove_page_podlove_settings_settings_handle
action:update
_wpnonce:f7c1a39d56
_wp_http_referer:/wp-admin/admin.php?page=podlove_settings_settings_handle&podlove_tab=web-player
podlove_webplayer_formats[audio][mp3]:1
podlove_webplayer_formats[audio][mp4]:0
podlove_webplayer_formats[audio][ogg]:2
podlove_webplayer_formats[audio][opus]:0
podlove_webplayer_formats[video][mp4]:0
podlove_webplayer_formats[video][ogg]:0
podlove_webplayer_formats[video][webm]:0
podlove_webplayer_settings[inject]:manually
podlove_webplayer_settings[chaptersVisible]:false
podlove_webplayer_settings[version]:player_v3
podlove_webplayer_settings[playerv3theme]:pwp-midnightblue.min.css
submit:Save Changes
```

**Comment by eteubert** [comment]
But then it wouldn't work for anyone. As far as I know you're currently the only one experiencing those problems, so I will need to replicate your environment when I find some spare time to investigate.

You may be right that it's whitelist related but the question then is _why_ it's not in the whitelist _in your environment_.

**Comment by blarghmatey** [comment]
@eteubert to replicate my environment you can clone this repository (https://gitlab.com/renaissancedev/podcastinit-salt), checkout the `podlove_debugging` branch and run `vagrant up`. Once it is complete you can log in as admin:admin at 192.168.33.10/wp-admin and install the podlove plugin. Once that is installed and activated you can attempt to edit any of the expert settings and save it and you should see the issue replicated.

Let me know if you need any help with getting that set up. Thank you for your time!

**Comment by blarghmatey** [comment]
@eteubert @timpritlove I just wanted to see if you received my last message and ask if there is any other information or help that I can provide.

**Comment by eteubert** [comment]
Got your message, and it's still on my radar. Just didn't have the time yet to try it out :-/

**Comment by blarghmatey** [comment]
I can understand and appreciate that. Thank you for getting back to me :)

**Comment by blarghmatey** [comment]
I tested this with PHP version 5.6 and experienced the same issue. I'm not sure what I should try next to debug this issue.

**Comment by eteubert** [comment]
Tried getting your box to run.

`http://192.168.33.10/wp-admin` shows me a piwik install page, which looks broken because all assets (like http://192.168.33.10/libs/jquery/themes/base/jquery-ui.min.css?cb=a22b62e99bb39c56aba223cbef8ef1b4) also point to the piwik install page. Skimming the provisioning log I can only see success messages, no errors. Help? :)

**Comment by eteubert** [comment]
Seems like it's a curious PHP but, at least according to http://stackoverflow.com/a/36205923/72448 which is what I can observe on your system:

> filter_input() doesn't work with INPUT_SERVER or INPUT_ENV when you use FASTCGI

The method that fails on your system is [is_options_save_page ](https://github.com/podlove/podlove-publisher/blob/397221f26848730c51872136882a9a0056e879ef/lib/helper.php#L94):

```php
function is_options_save_page() {
	$self    = filter_input(INPUT_SERVER, 'PHP_SELF');
	$request = filter_input(INPUT_SERVER, 'REQUEST_URI');

	return stripos($self, 'options.php') !== FALSE || stripos($request, 'options.php') !== FALSE;
}
```

However, I don't _need_ to use `filter_input` there, so I will just switch to `$_SERVER['PHP_SELF']` and `$_SERVER['REQUEST_URI']` here, which will solve the issue.

**Comment by eteubert** [comment]
Fix will be included in v2.4.

**Comment by blarghmatey** [comment]
I just activated the beta and tested and it works! Thank you for figuring this out and fixing it! :smile:
